packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp

```diff
@@ -398,55 +392,67 @@ class SerialDenseMatrix : public CompObject, public BLAS<OrdinalType, ScalarType
     ScalarType alpha = ScalarTraits<ScalarType>::zero() );
   void deleteArrays();
   void checkIndex( OrdinalType rowIndex, OrdinalType colIndex = 0 ) const;
-  OrdinalType numRows_;
-  OrdinalType numCols_;
-  OrdinalType stride_;
-  bool valuesCopied_;
-  ScalarType* values_;
 
+  static ScalarType*
+  allocateValues(const OrdinalType numRows,
+                 const OrdinalType numCols)
+  {
+    const size_t size = size_t(numRows) * size_t(numCols);
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wvla"
+    return new ScalarType[size];
+#pragma GCC diagnostic pop
+  }
+
+  OrdinalType numRows_ = 0;
+  OrdinalType numCols_ = 0;
+  OrdinalType stride_ = 0;
+  bool valuesCopied_ = false;
+  ScalarType* values_ = nullptr;
 }; // class Teuchos_SerialDenseMatrix
 
 //----------------------------------------------------------------------------------------------------
 //  Constructors and Destructor
 //----------------------------------------------------------------------------------------------------
 
-template<typename OrdinalType, typename ScalarType>
-SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix()
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(0), numCols_(0), stride_(0), valuesCopied_(false), values_(0)
-{}
-
 template<typename OrdinalType, typename ScalarType>
 SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   OrdinalType numRows_in, OrdinalType numCols_in, bool zeroOut
   )
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(numRows_in), numCols_(numCols_in), stride_(numRows_in)
+  : numRows_(numRows_in),
+    numCols_(numCols_in),
+    stride_(numRows_in),
+    valuesCopied_(true),
+    values_(allocateValues(numRows_in, numCols_in))
 {
-  values_ = new ScalarType[stride_*numCols_];
-  valuesCopied_ = true;
-  if (zeroOut == true)
+  if (zeroOut) {
     putScalar();
+  }
 }
 
 template<typename OrdinalType, typename ScalarType>
 SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   DataAccess CV, ScalarType* values_in, OrdinalType stride_in, OrdinalType numRows_in,
   OrdinalType numCols_in
   )
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(numRows_in), numCols_(numCols_in), stride_(stride_in),
-    valuesCopied_(false), values_(values_in)
+  : numRows_(numRows_in),
+    numCols_(numCols_in),
+    stride_(stride_in),
+    valuesCopied_(false),
+    values_(values_in)
 {
   if(CV == Copy)
   {
     stride_ = numRows_;
-    values_ = new ScalarType[stride_*numCols_];
+    values_ = allocateValues(stride_, numCols_);
     copyMat(values_in, stride_in, numRows_, numCols_, values_, stride_, 0, 0);
     valuesCopied_ = true;
   }
 }
 
 template<typename OrdinalType, typename ScalarType>
-SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseMatrix<OrdinalType, ScalarType> &Source, ETransp trans) 
-  : CompObject(),BLAS<OrdinalType,ScalarType>(), numRows_(0), numCols_(0), stride_(0), valuesCopied_(true), values_(0)
```

Comment:
And copy constructors need fewer explicit initializations with C++ in-class non-static member initializers :-)